File: js/src/engine/playing-sound.js
'use strict';

const [CONTEXT, RESOLVE, REJECT, NODES, NAME] = [Symbol(), Symbol(), Symbol(), Symbol(), Symbol()];

class PlayingSound extends Promise {
  constructor(context, nodes, name) {
    if(typeof context === 'function') {
      super(context);
      return;
    }
    let res, rej;
    super((resolve, reject) => {
      res = resolve;
      rej = reject;
      if(nodes.source) {
        nodes.source.addEventListener('ended', () => {
          this[RESOLVE]();
        });
      }
    });
    const cancel = () => {
      nodes.source.removeEventListener('ended', this[RESOLVE]);
      this[RESOLVE] = this[REJECT] = () => {};
    };
    this.then(cancel, cancel);
    this[RESOLVE] = res;
    this[REJECT] = rej;
    this[CONTEXT] = context;
    this[NODES] = nodes;
    this[NAME] = name;
  }

  get name() { return this[NAME]; }

  stop() {
    this[REJECT]();
    this[NODES].source && this[NODES].source.stop();
  }

  // the volume that a new instance of this sound will play at
  get volume() {
    if(this[NODES].gain) {
      return this[NODES].gain.gain.value;
    } else {
      return NaN;
    }
  }
  set volume(amt) {
    if(this[NODES].gain) {
      this[NODES].gain.gain.value = amt;
    }
  }
}

export default PlayingSound;

File: c++/src/engine/playing-sound.cc
#include "playing-sound.h"
#include "sound.h"

namespace Game {
    PlayingSound::PlayingSound(const Sound & sound) : _sound{ sound } {
        _channel = Mix_PlayChannel(-1, _sound._chunk.get(), 0);
        Mix_Volume(_channel, MIX_MAX_VOLUME);
    }

    void PlayingSound::stop() {
        Mix_HaltChannel(_channel);
    }

    float PlayingSound::volume() const {
        return static_cast<float>(Mix_Volume(_channel, -1)) / MIX_MAX_VOLUME;
    }

    void PlayingSound::volume(float amt) {
        Mix_Volume(_channel, static_cast<int>(amt * MIX_MAX_VOLUME));
    }
}
